src/remoteCollection/collect.js

```diff
@@ -232,68 +228,69 @@ function prepareRemoteRequest(generator) {
-      return generator;
+      return requestData;
     });
 } // prepareRemoteRequest
 
 /**
  * Collect data. Prepare, then send the remote request.
  *
- * @param  {Object} generator - The generator object
- * @returns {Promise} - resolves to a generator object with a "res"
- *  attribute carrying the response from the remote data source
+ * @param {Object} generator - The generator object
+ * @param {Array} subjects - The subjects to collect data for
+ * @returns {Promise} - resolves to a requestData object with attributes:
+ *  subjects, preparedUrl, preparedHeaders, res
  */
-function doCollect(generator) {
+function doCollect(generator, subjects) {
   debug('Entered "doCollect" for "%s"', generator.name);
-  return prepareRemoteRequest(generator)
-    .then(sendRemoteRequest);
+  return prepareRemoteRequest(generator, subjects)
+         .then((requestData) =>
+           sendRemoteRequest(generator, requestData)
+         );
 } // doCollect
 
 /**
  * Retrieves the list of subjects to collect data for, sets the array of
  * subjects into the generator, then prepares and executes the remote request.
  *
- * @param  {Object} generator - The generator object
- * @returns {Promise} - which resolves to a generator object with a "res"
- *  attribute carrying the response from the remote data source
+ * @param {Object} generator - The generator object
+ * @returns {Promise} - resolves to a requestData object with attributes:
+ *  subjects, preparedUrl, preparedHeaders, res
  * @throws {ValidationError} if thrown by prepareUrl (from sendRemoteRequest).
  */
 function collectBulk(generator) {
   debug('Entered "collectBulk" for "%s"', generator.name);
-  return attachSubjectsToGenerator(generator)
-    .then(doCollect);
+  return getSubjectsForGenerator(generator)
+         .then((subjects) =>
+           doCollect(generator, subjects)
+         );
 } // collectBulk
 
 /**
  * Retrieves the list of subjects to collect data for, sets the array of
  * subjects into the generator, then prepares and executes the remote request
  * for each subject.
  *
- * @param  {Object} generator - The generator object
- * @returns {Promise} - which resolves to an array of generator objects, each
- *  with a "res" attribute carrying the response from the remote data source.
+ * @param {Object} generator - The generator object
+ * @returns {Promise} - resolves to an array of requestData objects with attributes:
+ *  subjects, preparedUrl, preparedHeaders, res
  */
 function collectBySubject(generator) {
   debug('Entered "collectBySubject" for "%s"', generator.name);
-  return attachSubjectsToGenerator(generator)
-    .then((g) => bluebird.map(g.subjects, (subject) => {
-      // need to clone generator because we are doing async operation with
-      // generator data
-      const _g = JSON.parse(JSON.stringify(g));
-      _g.subjects = [subject];
-      return doCollect(_g);
-    }));
+  return getSubjectsForGenerator(generator)
+         .mapSeries((subject) =>
```

Comment:
Would this mean the collections will happen sequentially instead of parallel (like before)? If yes, could this have any side effect? What if a repeater process is not finished until next repeater starts? Just throwing my thoughts out there.

Reply:
Hmm, that's a good point, that would definitely be a plausible scenario when run sequentially if we had lots of subjects. We don't have any way of detecting or handling that case. Thinking this through, the next repeat cycle would start while the previous one is still finishing up, so the last few requests would be sent in parallel with the first few for the next cycle. The sample upserts aren't sent until all requests have completed, and the requests are tracked in-memory separately for each collection cycle, so the overlapping cycles wouldn't interfere with each other, and they would end up finishing and sending upserts a minute apart, as normal. So it would still work normally, just with some requests being sent in parallel, which is how it was before anyway. The only impact would be the offset from the repeat cycle would be greater. And probably greater memory usage because we would be tracking two collection cycles at once.

So I think if the repeater cycles overlapped it would be fine. However, just the fact that a collection could take a lot longer this way is maybe reason to reconsider.

I did it this way to make sure we only do the token request once. If we did it in parallel, when the token expires, the next collection cycle would send a separate auth request for each subject; this way it only sends it the first time. But maybe this is another case where it would be better to rethink the current approach than try and force the solution to fit it. 

The reason this is necessary in the first place is because the OAuth logic is implemented as part of sending the request. We could instead pull that out into a separate function, and run that once before sending any of the requests. I'm not sure exactly how that would work though since we currently rely on the result of the requests to tell us whether the token needs to be regenerated. Maybe when the token expires we just miss that cycle and request a new one next time? Or if any of the requests fail, we re-do all of them?

Reply:
Thank you both for thinking that through. Good question and good answer.

> The sample upserts aren't sent until all requests have completed

Does this mean that all the samples could be sitting and waiting because one of the requests might be slow and eventually times out at 30s, but with retries, everything else still waiting... ? So the rest of the samples don't get sent until that slow one finally times out again after all the retries and generates its error samples?

> Maybe when the token expires we just miss that cycle and request
> a new one next time?

IMO, missing a cycle is not an acceptable solution.

Reply:
Correct, they won't get sent until all the promises complete.

Reply:
Updated. 

I moved the token creation and request retries outside the request-specific code, and went back to making the requests in parallel. 

So now the by-subject collection flow looks like this:
- make a request to refocus to get subjects from subject query
- If necessary, make a request to the OAuth server to get a new token
- for each subject, in parallel, make a request to the data source
  - if any of the requests failed because of an expired token, retry the entire collection cycle, which will generate a new token
  - if any of the requests failed for other reasons, do nothing
- generate transform samples from the successful responses, or error samples from the failed ones

This way, all the asynchronous stuff happens outside the subject loop. (subject query, token creation, retries)